servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/IoExecutor.java

```diff
@@ -24,6 +24,15 @@
  */
 public interface IoExecutor extends ListenableAsyncCloseable {
 
+    /**
+     * Marker interface for IO Threads. All threads created by {@link IoExecutor}s are expected to implement this
+     * interface.
+     *
+     * @see #isIoExecutorThread(Thread)
+     */
+    interface IoThread {
```

Comment:
My biggest worry is that there is no way for users to use `IoThreadFactory`, because it's an internal class. Anyone, who is creating a custom `IoExecutor` (I'm sure there are such users), will end up being non-offloaded. 

Did you consider the approach we discussed offline of adding `IoThreadFactory` interface in `transport-api`?
```java
    interface IoThreadFactory<T extends Thread & IoThread> extends ThreadFactory {

        @Override
        T newThread(Runnable r);
    }
``` 
Then changing `io.servicetalk.transport.netty.NettyIoExecutors` to require it.

Ideally, we need to introduce `IoThreadFactory` and and methods for `NettyIoExecutors` in 0.41 branch. Then in `main` branch we can remove old static factories and make sure we won't break users.

Reply:
This does seem like a good option.

Reply:
It seems like this might only be possible as a hard breaking change since there will be no graceful support for IO threads that do not implement IoThread marker, ServiceTalk would just break if you tried it because necessary offloading would not be done.

It does seem to be the right answer though.

Reply:
Discussed offline. We will introduce `IoThreadFactory` interface in 0.41, add new factory methods for `NettyIoExecutors`, and deprecate old factories. Then we can break API in 0.42+

Reply:
Leaving this open until PR for 0.41 branch is created.

Reply:
Deprecation PR now available as #1738